### ISSUE TITLE
No hand animation when using Launch game

### ISSUE THREAD

**Issue by joticarroll** [ValveSoftware/steamvr_unreal_plugin#151]
**Affected Versions:** All versions

**Steps to reproduce:**
1.) Download SteamVRInputPlugin_UEIntegrated example project
2.) Right click the project file and select "Launch game" in windows file explorer

**Results:** Hands animation blueprint fails to load and hands are left at there origin un-animated.

**Expected Result:** Hand animation matching the PIE and published result.

**Crash?:** No

**Comment by 1runeberg** [comment]
Hi Joti! Can you post please the log file from the test project?

Thanks,
Rune

**Comment by 1runeberg** [comment]
Please note if you are using an Oculus headset, to disable the Oculus plugin - tracking and input should work in an uncooked build (right click > Launch). So check the LogHMD entry to determine which plugin took responsibility for Input or yeah post the log and I can check for any other issues.

As for the animation itself, my understanding is that dynamically loaded UE Animations (as was used in the sample project) need to be cooked so you need to run them from a packaged build or via the Editor.

**Comment by joticarroll** [comment]
[SteamVRInputPlugin.log](https://github.com/ValveSoftware/steamvr_unreal_plugin/files/4795751/SteamVRInputPlugin.log)

**Comment by joticarroll** [comment]
I had disabled the Oculus plugin already, this has been a problem with all previous versions of SteamVR Inputs I just didn't get around to reporting it until now.  Animation blueprints normally run fine in uncooked content and launch game mode is essential for good testing workflows, especially when building multi-user systems in a team.  Network sub-systems are not loaded in the editor and PIE sessions have altered map names making even LAN sessions a problem.  So the only reasonable approach is to use uncooked editor -game based launching of the project or maps.

**Comment by joticarroll** [comment]
`{
			"Name": "SteamVREditor",
			"Type": "Editor",
			"LoadingPhase": "PostEngineInit",
			"WhitelistPlatforms": [
				"Win64",
				"Win32",
				"Linux",
				"Mac"
			]
		}`
"UnrealEngine\Engine\Plugins\Runtime\Steam\SteamVR\SteamVR.uplugin"
It maybe as simple as changing the type var to "UnCookedOnly" in the plugin definition but I haven't had time to test and I'm trying stay away from custom source builds at the moment.

**Comment by 1runeberg** [comment]
The issue is around custom animation nodes. To create them, part of the code exists in an editor only module, which UE doesnt appear to load in an uncooked launch although techinically its an editor session, Hence you get the following:

[2020.06.18-01.10.54:865][  0]LogUObjectGlobals: Warning: While loading '../../../../../../Unreal/projects_svn/SteamVRInputPlugin_UEIntegrated/Content/Sample/Characters/Hands/SteamVR_Animations/ANIM_Glove_Right.uasset' failed to load '/Script/SteamVREditor': Can't find file.

Lookign at your logs, it seems the trackign is registering properly - can you confirm that the hands arent tracking and there is no input at all when you're using Rift S? I just tested with both Touch & Rift S, and could repro the tracking part of the issue.

**Comment by 1runeberg** [comment]
The issue is around custom animation nodes. To create them, part of the code exists in an editor only module to expose it to Blueprint, which UE doesnt appear to dynamically load in an uncooked launch although techinically its an editor session, Hence you get the following:

[2020.06.18-01.10.54:865][ 0]LogUObjectGlobals: Warning: While loading '../../../../../../Unreal/projects_svn/SteamVRInputPlugin_UEIntegrated/Content/Sample/Characters/Hands/SteamVR_Animations/ANIM_Glove_Right.uasset' failed to load '/Script/SteamVREditor': Can't find file.

Looking at your logs, it seems the tracking is registering properly - can you confirm that the hands arent tracking and there is no input at all when you're using Rift S? I just tested with both Touch & Rift S, and could not repro the tracking part of the issue. or am I misreading that part?

**Comment by joticarroll** [comment]
That's correct it's just the animation nodes that are failing, although they prevent the entire animation blueprint from operating as a side effect.  Input mapping still operates properly, although I have a number of issues I've discovered on that front too that I'll put in a separate issue entry.

**Comment by joticarroll** [comment]
No I haven't had any tracking or input issues in relation to this problem.

**Comment by 1runeberg** [comment]
> That's correct it's just the animation nodes that are failing, although they prevent the entire animation blueprint from operating as a side effect. Input mapping still operates properly, although I have a number of issues I've discovered on that front too that I'll put in a separate issue entry.

i dont think we have any control on this from the plugin side however as this is an engine behavior. nor can we move the module to the runtime. basically, the engine needs to dynamically load the animation nodes meant for editor sessions into their uncooked launch mode - i suggest opening a separate issue in UE for this, and feel free to use our plugin as an example of the behavior. will poke epic internally, but its always better to have direct reports from developers.

on our side, we're looking at putting back livelink support as part of  our openxr efforts and augment any current implementation limitations in that area with steamvr skeletal animations.

thanks for confirming the tracking part - and yes please, any other input issues, please report on separate issues and would be glad to investigate and fix anything that we can from our end.

**Comment by joticarroll** [comment]
Thanks for looking into it Rune.

**Comment by 1runeberg** [comment]
ah... it seems Epic has addressed this already. The plugin came out around the very early days of LiveLink so the animation schema is based off that (though LiveLink support was later removed to support older engine versions - from 4.15).

my bad i havent kept up, they've recently renamed the Developer module to UncookedOnly and that's now best practice to hold animnodes so -game "editor" launches can have access to these anim nodes.

will introduce a new module in 4.26, and make sure the LiveLink version also does the same. thanks for the report Joti! 

Will keep this open until our PR is fully merged and 4.26 released.

**Comment by equalent** [comment]
Hi!

Similar problem here: ABPs with SteamVR Skeletal Anim Pose work correctly but SteamVR nodes are removed from the graph after the engine is restarted (with and without the editor).

Log:
```
LogLinker: Warning: While loading '../../../../kurushimi/KuGame/Content/KuGame/Character/ViewModel/ABP_HandL.uasset' failed to load '/Script/SteamVREditor': Can't find file.
LogLinker: Warning: While loading '../../../../kurushimi/KuGame/Content/KuGame/Character/ViewModel/ABP_HandL.uasset' failed to load '/Script/SteamVREditor': Can't find file.
LogLinker: Warning: While loading '../../../../kurushimi/KuGame/Content/KuGame/Character/ViewModel/ABP_HandL.uasset' failed to load '/Script/SteamVREditor': Can't find file.
LogLinker: Warning: While loading '../../../../kurushimi/KuGame/Content/KuGame/Character/ViewModel/ABP_HandL.uasset' failed to load '/Script/SteamVREditor': Can't find file.
```
